packages/sync/src/queue/index.ts

```diff
@@ -0,0 +1,64 @@
+import PQueue from 'p-queue'
+import { ConsecutiveList } from './consecutive-list'
+
+export { ConsecutiveList }
+
+export type SyncQueueOptions = {
+  setCursor?: (cursor: number) => Promise<void>
+  concurrency?: number
+  startCursor?: number
+}
+
+// A queue with arbitrarily many partitions, each processing work sequentially.
+// Partitions are created lazily and taken out of memory when they go idle.
+export class SyncQueue {
+  consecutive = new ConsecutiveList<number>()
+  mainQueue: PQueue
+  partitions = new Map<string, PQueue>()
+  cursor: number
+
+  constructor(public opts: SyncQueueOptions = {}) {
+    this.mainQueue = new PQueue({ concurrency: opts.concurrency ?? Infinity })
+    this.cursor = opts.startCursor ?? 0
```

Comment:
Is it more natural to backfill or to hop right into the stream by default?  I lean towards the latter because it's less of a commitment, less load on the relay, and both are fairly arbitrary times to pick-up from.

Reply:
yeah you're right 👍